### ISSUE TITLE
Regarding `JCBC-2152` 

### ISSUE THREAD

**Issue by yeikel** [couchbase/couchbase-jvm-clients#29]
Hi team,

I was hoping to collaborate on [JCBC-2152](https://jira.issues.couchbase.com/browse/JCBC-2152) but sadly the new jira is readonly for external users 

If it helps in any way, this bug was introduced with `3.6.1`. In `3.6.0` this does not happen

Thanks! 

[JCBC-2152]: https://couchbasecloud.atlassian.net/browse/JCBC-2152?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by mikereiche** [comment]
Thanks.  I've noted it on the jira ticket.

**Comment by yeikel** [comment]
We suspect that this is the commit that introduced the problem https://github.com/couchbase/couchbase-jvm-clients/commit/cf37f2f39c268c8f83ff40c21e00c02ddb3a42f2

**Comment by yeikel** [comment]
Hi team,

Is there any ETA for this issue?

Unfortunately, we're locked to an older version until this is fixed, and we're missing many bug fixes and vulnerability changes as a result 

This issue is a considerable breaking change, and it is unfortunate that it is not prioritized

If this is the new expected behavior, please clarify

Thank you

**Comment by programmatix** [comment]
Hi @yeikel.  This one's on my queue.  I'm not debating that this is a regression, and I apologise that you've been impacted by it - we certainly will get it fixed, and appreciate you reporting it.  But from our side, it seems a relatively minor regression, and this is why it hasn't been prioritised ahead of other work: generally there is no need to have the updated TransactionGetResult, because there is nothing particularly useful you can do with it.  Could you please let me know why this is a blocking problem for you?  Would be good to understand your use-case better.

**Comment by ianprime0509** [comment]
Thank you for the update! Just to add more information here (I work on the same team as Yeikel): in our platform, we've written a thin wrapper around the Couchbase SDK for internal use that provides integration with some other internal components, logging, etc. The API we've written around transactions is intended to support the following pattern:

```java
couchbaseService
    .executeTransaction(ctx -> ctx.get(key)
        .flatMap(result -> ctx.replace(result, newContent))
        .map(GetResult::document))
    .flatMap(updatedDocument -> /* perform additional logic with the updated document content */);
```

(`executeTransaction` resolves with the value returned by its "action" function argument once the transaction completes successfully) The property that `updatedDocument` contains the updated document content in this example is what one of our internal platform test cases is testing for, which fails with newer SDK versions (the reason we haven't upgraded yet).

It's not impossible for us to work around this: since these are all wrapper methods, we can update our wrapper implementation of `ctx.replace` to ignore what the Couchbase SDK provides as the document content and use (a copy of) `newContent` instead. One pattern I was worried would not work with this workaround is the following adaptation of the test case in JCBC-2152:

```java
@Test
void testTransaction() {
  String docId = UUID.randomUUID().toString();
  JsonObject initial = JsonObject.create().put("foo", "bar");
  JsonObject updated = JsonObject.create().put("foo", "baz");
  JsonObject updated2 = JsonObject.create().put("foo", "baz2");
  collection.insert(docId, initial);

  cluster.transactions().run((ctx) -> {
    TransactionGetResult oldDoc = ctx.get(collection, docId);
    TransactionGetResult replacedDoc = ctx.replace(oldDoc, updated);
    TransactionGetResult replacedDoc2 = ctx.replace(replacedDoc, updated2);
    TransactionGetResult ryowDoc = ctx.get(collection, docId);

    LOGGER.info("Original document: {}", oldDoc.contentAsObject());
    LOGGER.info("Replaced document: {}", replacedDoc.contentAsObject());
    LOGGER.info("Replaced document 2: {}", replacedDoc2.contentAsObject());
    LOGGER.info("Final document: {}", ryowDoc.contentAsObject());
  });
}
```

My concern was that, since `replacedDoc` does not have the updated document content (and presumably metadata), the call `ctx.replace(replacedDoc, updated2)` might not work, since it was not clear whether there were internal consistency checks on the `TransactionGetResult` passed as the first argument which would enforce the most recent "revision" of the document to be passed. However, this test case seems to work fine even with newer SDK versions, so I don't see any major impediment to the workaround mentioned above.

I think the motivation behind Yeikel's comment is that if a fix for this within the SDK is coming soon, then we can just update to the fixed version and avoid applying this workaround in our wrapper. However, given the fix may be further in the future than we anticipated, I personally don't see any issue with applying the workaround for now and revisiting it when the linked issue has been resolved.

**Comment by programmatix** [comment]
Hi @ianprime0509 thank you for the detailed explanation, I think I get it.  And as I mentioned and you confirm, the result of ctx.replace() is essentially giving you what you already have access to (`newContent`), and so the workaround you mention is available to you.  But nonetheless, let me see what I can do to prioritise getting the fix into the next SDK release (ETA: early Jan).

**Comment by programmatix** [comment]
This is now fixed and will be in the 3.7.7 Java SDK release, ETA is early January.
